sdk/identity/azure-identity/src/test/java/com/azure/identity/AzurePipelinesCredentialTest.java

```diff
@@ -44,12 +54,29 @@ public void testGetTokenFromPipeline() {
     }
 
     @Test
-    @LiveOnly
     public void testGetTokenFromPipelineSync() {
         // Arrange
         AzurePipelinesCredential credential = getCredential();
 
         // Act & Assert
         assertNotNull(credential.getTokenSync(new TokenRequestContext().addScopes("https://vault.azure.net/.default")));
     }
+
+    @Test
+    public void noRedirectOnFailure() {
+        AzurePipelinesCredential credential = getInvalidCredential();
+        StepVerifier.create(credential.getToken(new TokenRequestContext().addScopes("https://vault.azure.net/.default")))
+            .expectErrorMatches(throwable ->
+                ((ClientAuthenticationException) throwable.getCause()).getResponse().getStatusCode() == 401
+            ).verify();
+    }
+
+    @Test
+    public void validateHeadersInException() {
+        AzurePipelinesCredential credential = getInvalidCredential();
+        StepVerifier.create(credential.getToken(new TokenRequestContext().addScopes("https://vault.azure.net/.default")))
+            .expectErrorMatches(throwable ->
+                throwable.getMessage().contains("x-vss-e2eid") && throwable.getMessage().contains("x-msedge-ref")
+            ).verify();
```

Comment:
I suspect this test will pass, live, as expected since the service will return these two header values in the response (in this case). It's a good test to add.

However, it is possible that for some other case, one or both of these don't come back in the response. Let's keep an eye out if we see the expected response headers to be missing, causing the test to fail. 

In C++, we added "non-live" tests with mocked responses with one or both of the headers missing to exercise those branches and make sure the exception throwing logic doesn't break in an unexpected way, for that case. Up to you to consider adding.

Reply:
Given that missing headers will just result in nulls, which will then just not add that to the exception in the case it happens in the field, I think what I've got here is fine.